src/xmipp/libraries/reconstruction/resolution_fso.cpp

```diff
@@ -0,0 +1,1876 @@
+    	}
+
+
+//    	filteredMap.resizeNoCopy(half1);
+    	filteredMap.resizeNoCopy(m1sizeX, m1sizeY, m1sizeZ);
+    	transformer1.inverseFourierTransform(FThalf1, filteredMap);
+    }
+
+
+    void resolutionDistribution(MultidimArray<double> &resDirFSC, FileName &fn)
+    {
+    	Matrix2D<int> anglesResolution;
+    	size_t Nrot = 360;
+    	size_t Ntilt = 91;
+    	size_t objIdOut;
+
+    	MetaData mdOut;
+    	Matrix2D<double> w, wt;
+    	w.initZeros(Nrot, Ntilt);
+    	wt = w;
+    	double cosAngle = cos(ang_con);
+    	double aux = 4.0/((cosAngle -1)*(cosAngle -1));
+    	// Directional resolution is store in a metadata
+    	for (double i=0; i<Nrot; i++)
+		{
+			for (double j=0; j<Ntilt; j++)
+			{
+				double rotmatrix =  i*PI/180;
+				double tiltmatrix = j*PI/180;
+				double xx = sin(tiltmatrix)*cos(rotmatrix);
+				double yy = sin(tiltmatrix)*sin(rotmatrix);
+				double zz = cos(tiltmatrix);
+
+				double w = 0;
+				double wt = 0;
+				for (size_t k = 0; k<angles.mdimx; k++)
+				{
+					double rot = MAT_ELEM(angles, 0, k);
+					double tilt = MAT_ELEM(angles, 1, k);
+
+					rot *= PI/180;
+					tilt *= PI/180;
+
+					double x_dir = sin(tilt)*cos(rot);
+					double y_dir = sin(tilt)*sin(rot);
+					double z_dir = cos(tilt);
+
+					double cosine = fabs(x_dir*xx + y_dir*yy + z_dir*zz);
+					if (cosine>=cosAngle)
+					{
+						cosine = exp( -((cosine -1)*(cosine -1))*aux );
+						w += cosine*( dAi(resDirFSC, k) );
+						wt += cosine;
+					}
+				}
+
+	    	double wRes = w/wt;
+			objIdOut = mdOut.addObject();
+			mdOut.setValue(MDL_ANGLE_ROT, i, objIdOut);
+			mdOut.setValue(MDL_ANGLE_TILT, j, objIdOut);
+			mdOut.setValue(MDL_RESOLUTION_FRC, wRes, objIdOut);
+			}
+		}
+		mdOut.write(fn);
+    }
+
+
+
+    void getCompleteFourier(MultidimArray<double> &V, MultidimArray<double> &newV,
+    		int m1sizeX, int m1sizeY, int m1sizeZ)
+        {
+    	newV.resizeNoCopy(m1sizeX, m1sizeY, m1sizeZ);
+		int ndim=3;
+		if (m1sizeX==1)
+		{
+			ndim=2;
+			if (m1sizeY==1)
+				ndim=1;
+		}
+		double *ptrSource=NULL;
+		double *ptrDest=NULL;
+		FOR_ALL_DIRECT_ELEMENTS_IN_ARRAY3D(newV)
+		{
+			ptrDest=(double*)&DIRECT_A3D_ELEM(newV,k,i,j);
+			if (j<XSIZE(V))
+			{
+				ptrSource=(double*)&DIRECT_A3D_ELEM(V,k,i,j);
+				*ptrDest=*ptrSource;
+//				*(ptrDest+1)=*(ptrSource+1);
+			}
+			else
+			{
+				ptrSource=(double*)&DIRECT_A3D_ELEM(V,
+													(m1sizeZ-k)%m1sizeZ,
+													(m1sizeY-i)%m1sizeY,
+													m1sizeX-j);
+				*ptrDest=*ptrSource;
+//				*(ptrDest+1)=-(*(ptrSource+1));
+			}
+		}
+   }
+
+
+
+
+    void createFullFourier(MultidimArray<double> &fourierHalf, FileName &fnMap,
+    		int m1sizeX, int m1sizeY, int m1sizeZ)
+    {
+    	MultidimArray<double> fullMap;
+		getCompleteFourier(fourierHalf, fullMap, m1sizeX, m1sizeY, m1sizeZ);
+		CenterFFT(fullMap, true);
+		Image<double> saveImg;
+		saveImg() = fullMap;
+	    saveImg.write(fnMap);
+    }
+
+    void run()
```

Comment:
very long method

Reply:
This is an almost final version, I still need a lot of functions that I cannot delete now